library/arith.lisp

```diff
@@ -502,6 +508,159 @@ The fields are defined as follows:
+      ((tuple (EQ) (GT)) (/ pi (fromint 2)))
+      ((tuple (EQ) (LT)) (/ pi (fromint -2)))
+      ((tuple (EQ) (EQ)) 0))))
+
+
+(cl:defmacro %define-real-float-instance
+    (coalton-type cl:&key
+     (underlying-type (cl:find-symbol (cl:symbol-name coalton-type) :cl)))
+  "Defines the float instances for a lisp floating-point type"
+  `(coalton-toplevel
+     (define-instance (Trigonometric ,coalton-type)
+       ,(generate-unary-wrapper coalton-type 'sin)
+       ,(generate-unary-wrapper coalton-type 'cos)
+       ,(generate-unary-wrapper coalton-type 'tan)
+       ,(generate-unary-wrapper coalton-type 'asin :domain '(lambda (x) (cl:<= -1 x 1)))
+       ,(generate-unary-wrapper coalton-type 'acos :domain '(lambda (x) (cl:<= -1 x 1)))
+       ,(generate-unary-wrapper coalton-type 'atan))
+
+     (define-instance (Exponentiable ,coalton-type)
+       (define (expt base power)
+         (lisp ,coalton-type (base power)
+           (cl:let ((res (cl:expt base power)))
+             (cl:when (cl:complexp res)
+               (cl:error "The exponentiation ~a ^ ~a is not real" base power))
+             res)))
+       (define (log base number)
+         (lisp ,coalton-type (base number)
+           (cl:when (cl:<= base 0)
+             (cl:error "Cannot take a logarithm with the non-positive real base ~a" base))
+           (cl:when (cl:<= number 0)
+             (cl:error "Cannot take the logarithm of the non-positive real number ~a" number))
+           (cl:log number base))))
+
+     (define-instance (Into Fraction ,coalton-type)
+       (define (into x)
+         (lisp ,coalton-type (x)
+           (cl:coerce x ',underlying-type))))
+     
+     (define-instance (Float ,coalton-type)
+       (define ee
+         (lisp ,coalton-type ()
+           (cl:exp (cl:coerce 1 ',underlying-type))))
+         
+       (define pi
+         (lisp ,coalton-type ()
+           (cl:coerce cl:pi ',underlying-type))))
+
+     (define-instance (RealFloat ,coalton-type)
+       (define (rationalize x)
+         (lisp fraction (x)
+           (cl:coerce (cl:rationalize x) 'cl:ratio))))))
+
+(%define-real-float-instance single-float)
+(%define-real-float-instance double-float)
+
+(cl:defmacro %define-integer-expt-instance (coalton-type)
+  `(coalton-toplevel
+     (define-instance (Exponentiable ,coalton-type)
+       (define (expt base power)
+         (lisp ,coalton-type (base power)
+           (cl:when (cl:< power 0)
+             (cl:error "Cannot exponentiate to the power of the negative integer ~a" power))
+           (cl:expt base power)))
+       (define (log base number)
+         (lisp ,coalton-type (base number)
+           (cl:when (cl:<= base 0)
+             (cl:error "Cannot take a logarithm with the non-positive real base ~a" base))
+           (cl:when (cl:<= number 0)
+             (cl:error "Cannot take the logarithm of the non-positive real number ~a" number))
+           (cl:values (cl:floor (cl:log number base))))))))
```

Comment:
this will be error prone for large values... we'll need a more clever algorithm (probably using integer-length somehow)

Reply:
I've implemented Aidan's algorithm